packages/belos/tpetra/test/BlockGmres/test_bl_fgmres_hb.cpp

```diff
@@ -0,0 +1,271 @@
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+//
+// This driver reads a problem from a Harwell-Boeing (HB) file.
+// The right-hand-side corresponds to a randomly generated solution.
+// The initial guesses are all set to zero.
+//
+// NOTE: No preconditioner is used in this case.
+//
+#include "BelosConfigDefs.hpp"
+#include "BelosLinearProblem.hpp"
+#include "BelosTpetraAdapter.hpp"
+#include "BelosTpetraOperator.hpp"
+#include "BelosBlockGmresSolMgr.hpp"
+
+// I/O for Harwell-Boeing files
+#define HIDE_TPETRA_INOUT_IMPLEMENTATIONS
+#include <Tpetra_MatrixIO.hpp>
+
+#include <Teuchos_CommandLineProcessor.hpp>
+#include <Teuchos_ParameterList.hpp>
+#include <Teuchos_StandardCatchMacros.hpp>
+#include <Tpetra_Core.hpp>
+#include <Tpetra_CrsMatrix.hpp>
+
+using namespace Teuchos;
+using Tpetra::Operator;
+using Tpetra::CrsMatrix;
+using Tpetra::MultiVector;
+using std::endl;
+using std::cout;
+using std::vector;
+using Teuchos::tuple;
+
+int main(int argc, char *argv[]) {
+
+  typedef double                           ST;
+  typedef ScalarTraits<ST>                SCT;
+  typedef SCT::magnitudeType               MT;
+  typedef Tpetra::Operator<ST>             OP;
+  typedef Tpetra::MultiVector<ST>          MV;
+  typedef Belos::OperatorTraits<ST,MV,OP> OPT;
+  typedef Belos::MultiVecTraits<ST,MV>    MVT;
+
+  Tpetra::ScopeGuard tpetraScope(&argc,&argv);
+
+  bool success = false;
+  bool verbose = false;
+  try {
+    const ST one  = SCT::one();
+
+    int MyPID = 0;
+
+    RCP<const Comm<int> > comm = Tpetra::getDefaultComm();
+    //
+    // Get test parameters from command-line processor
+    //
+    bool proc_verbose = false;
+    bool debug = false;
+    int frequency = -1;  // how often residuals are printed by solver
+    int numrhs = 1;      // total number of right-hand sides to solve for
+    int blocksize = 1;   // blocksize used by solver
+    int maxiters = -1;   // maximum number of iterations for solver to use
+    int length = 100;    // max subspace size
+    std::string filename("bcsstk14.hb");
+    MT tol = 1.0e-5;     // relative residual tolerance
```

Comment:
The default should depend on the Scalar type, e.g., via `Teuchos::ScalarTraits<Scalar>::eps()`.

Reply:
This would make convergence more dependent on scalar type, but I haven't seen such a convention used in any other Belos tests.

Reply:
@jennloe If "Scalar" is always `double` then it doesn't matter so much.  If you intend for the test to be generic, then it matters.  In particular, `Scalar=float` does not get enabled by default in PR tests (even though Sierra depends on it! @trilinos/framework ), so if you intend to test that case, then the tolerance should be a function of the Scalar type.

Reply:
I'm going to create a separate issue for this.  It affects nearly every Belos test, and I am not ready to change it here.

Reply:
Or maybe I won't.  For these particular block GMRES tests, the scalar type is hard-coded as double.  It would need further work to generalize.  
There are some tests for block GMRES and CG that test both double and float.

Reply:
In other concerns- Is there a reason that Tpetra doesn't have tests for BiCGStab, TFQMR, or MinRes?  Or PseudoBlockGMRES?  We might be asking for trouble when Epetra gets phased out.

Reply:
@jennloe There is no good reason for there not being tests for the other Belos solvers and you bring up a good point that this code coverage is essential to ensuring correctness between Tpetra and Belos.  I say that you move forward with the focus of this particular pull request.  Addressing further necessary improvements to testing is beyond the scope of this work.